libs/@local/harpc/net/src/session/server/connection/collection.rs

```diff
@@ -0,0 +1,383 @@
+        // explicitely here, as to not leave any dangling tasks.
+        self.cancel_all();
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct TransactionPermit {
+    id: RequestId,
+    // adding an overflowing generation counter helps to prevent us from accidentally
+    // removing requests in the case that we're overriding them.
+    // If we override, we *could* have the case where w/ the following code:
+    //
+    // 1. TransactionCollection::acquire()
+    //    - create a new permit
+    //    - cancel the old permit (this one)
+    //    - replace old state (this one is bound to) with new state
+    // 2. TransactionPermit::drop() (old)
+    //    - remove the state from the storage
+    //
+    // without the generation, we would, in that case, remove the new state, which is not
+    // what we want.
+    //
+    // Also known as the ABA problem.
+    generation: u64,
+
+    storage: TransactionStorage,
+
+    cancel: CancellationToken,
+    notify: Arc<Notify>,
+
+    _permit: ConcurrencyPermit,
+}
+
+impl TransactionPermit {
+    fn new(
+        collection: &TransactionCollection,
+        id: RequestId,
+        cancel: CancellationToken,
+    ) -> Result<Arc<Self>, ConnectionTransactionLimitReachedError> {
+        let permit = collection.limit.acquire()?;
+        let storage = Arc::clone(&collection.storage);
+        let notify = Arc::clone(&collection.notify);
+
+        // we only need to increment the generation counter, when we are successful in acquiring a
+        // permit.
+        // Relaxed ordering is fine here, as we only ever increment the value and we do not care
+        // about the order of those increments.
+        let generation = collection.generation.fetch_add(1, Ordering::Relaxed);
+
+        Ok(Arc::new(Self {
+            id,
+            storage,
+            generation,
+            notify,
+            _permit: permit,
+            cancel,
+        }))
+    }
+}
+
+impl ServerTransactionPermit for TransactionPermit {
+    fn cancellation_token(&self) -> &CancellationToken {
+        &self.cancel
+    }
+
+    fn id(&self) -> RequestId {
+        self.id
+    }
+}
+
+impl Drop for TransactionPermit {
+    fn drop(&mut self) {
+        // using async and sync methods can lead to deadlocks, so we spawn a new task to remove the
+        // state from the storage (it's also just cleaner this way), as `remove_if` might block.
+        let storage = Arc::clone(&self.storage);
+        let notify = Arc::clone(&self.notify);
+
+        let id = self.id;
+        let generation = self.generation;
+
+        tokio::spawn(async move {
+            let has_removed = storage
+                .remove_if_async(&id, |state| state.generation == generation)
+                .await;
+
+            // we only need to check if empty, if we have removed something
+            if has_removed && storage.is_empty() {
+                notify.notify_one();
+            }
+        });
+    }
```

Comment:
Similar to the client code: it feels wild to spawn a thread in `Drop`.

Reply:
added explanation in [7a79dba](https://github.com/hashintel/hash/pull/4459/commits/7a79dbae5a36851a976933d9646bcbec4af143d6)